File: easy/array/x_card_partition/x_card_partition.go
package main

import "fmt"

func gcd(a, b int) int {
	if b == 0 {
		return a
	}
	return gcd(b, a%b)
}

func hasGroupsSizeX(deck []int) bool {
	numberToOccurence := make(map[int]int)
	for _, c := range deck {
		if _, ok := numberToOccurence[c]; ok {
			numberToOccurence[c]++
		} else {
			numberToOccurence[c] = 1
		}
	}

	x := numberToOccurence[deck[0]]
	if x < 2 {
		return false
	}

	for _, occurence := range numberToOccurence {
		if gcd(x, occurence) < 2 {
			return false
		}
		x = gcd(x, occurence)
	}
	return true
}

func main() {
	fmt.Println("Solution 1 : ", hasGroupsSizeX([]int{1, 2, 3, 4, 4, 3, 2, 1}))
	fmt.Println("Solution 2 : ", hasGroupsSizeX([]int{1, 1, 1, 2, 2, 2, 3, 3}))
	fmt.Println("Solution 3 : ", hasGroupsSizeX([]int{1}))
	fmt.Println("Solution 4 : ", hasGroupsSizeX([]int{1, 1}))
	fmt.Println("Solution 5 : ", hasGroupsSizeX([]int{1, 1, 2, 2, 2, 2}))
}

File: easy/array/x_card_partition/x_card_partition.py
class Solution(object):
    def __gcd(self, a, b):
        if a == 0:
            return b
        return self.__gcd(b % a, a)

    def hasGroupsSizeX(self, deck):
        """
        :type deck: List[int]
        :rtype: bool
        """
        occurrences = {}
        for c in deck:
            if not c in occurrences:
                occurrences[c] = deck.count(c)

        x = occurrences[deck[0]]
        for _, o in occurrences.items():
            if self.__gcd(x, o) < 2:
                return False
            x = self.__gcd(x, o)

        return True


s = Solution()
print("Solution 1 : ", s.hasGroupsSizeX([1, 2, 3, 4, 4, 3, 2, 1]))
print("Solution 2 : ", s.hasGroupsSizeX([1, 1, 1, 2, 2, 2, 3, 3]))
print("Solution 3 : ", s.hasGroupsSizeX([1]))
print("Solution 4 : ", s.hasGroupsSizeX([1, 1]))
print("Solution 5 : ", s.hasGroupsSizeX([1, 1, 2, 2, 2, 2]))
